### DISCUSSION TITLE
Support for multifactor authentication

### DISCUSSION THREAD

**Discussion by maloo1** [schwabe/ics-openvpn#1490] [Configuration problems]
Hi,
I was was wondering if you support multifactor authentication - ie: a username/password and One time passcode (OTP) as mentioned here: https://github.com/evgeny-gridasov/openvpn-otp#using-openvpn-otp-for-multi-factor-authentication

Searching the Issues I see you do not support "static-challenge" but it sounded like there was another way to achieve this??

Thanks in advance!

**Comment by schwabe** [comment]
The app supports the more modern auth pending mechanism (see here https://github.com/OpenVPN/openvpn/blob/master/doc/management-notes.txt).

**Reply by maloo1** [reply]
Thanks Arne, and thanks for a great app!

Do you know how I would configure you app to provide both a username/password and OTP token? I see that text document you linked does support the "static-challenge" but I'm just not sure how to enter this when connecting??

**Reply by schwabe** [reply]
The static-challenge described in that document is not supported only the pending auth method documented in that document is.

**Reply by maloo1** [reply]
Thanks Arne, Do you know if your planning on implementing it or is there another way to use an OTP with openVPN??

**Reply by schwabe** [reply]
The authpending method is already used for OTP/2FA with OpenVPN Access Server.

You need to use the management interface or those two patches https://patchwork.openvpn.net/patch/1899/ and https://patchwork.openvpn.net/patch/1816/

**Reply by schwabe** [reply]
But regarding static-challenge support. No, I don't plan on implementing it.

**Reply by maloo1** [reply]
Oh ok nice, I assume those patches need to be applied to the community version of openvpn? Do you know if/when they are planning on merging this (if not already merged)?

Can "client-crresponse" be implemented along with the plugin used by openvpn-auth-ldap to achieve MFA? Or is there a another way to achieve that?

FYI I'm just trying to implement MFA without using the access server. I have this working via the "community" version of the windows client but cannot connect from an android device...

**Reply by schwabe** [reply]
They should be merged before the release of OpenVPN 2.6. I personally do not have experience with  openvpn-auth-ldap plugin but they should work together.

**Reply by mcepl** [reply]
How can I find out what type of server our company uses, and why (assuming we use the latest version) I am still getting this error when trying to connect:

```
01:46 AUTH: Received control message: AUTH_FAILED,CRV1:R,E:something=:somethinglese=:Enter Your OTP Code
```

**Reply by schwabe** [reply]
ask your administrator what kind of server you are using.

**Reply by mcepl** [reply]
Regular FLOSS [OpenVPN from SUSE Enterprise Linux](https://build.opensuse.org/package/show/SUSE:SLE-15-SP6:Update/openvpn), the response is from this plugin we use https://github.com/SUSE/openvpn-mfa-plugin/blob/295e4efe84564b67844e87e7b816e50038283e92/src/plugin.rs#L209

**Reply by schwabe** [reply]
Yeah. It is unfortunate that even though that plugin was apparently written this year, is only supporting the old/legacy MFA functionality.

**Reply by mcepl** [reply]
This is what I’ve got from our people:

> We briefly touched this "dynamic challenge" (AUTH_PENDING) they are mentioning when designing this solution.
> We need to support a wide range of clients - Windows OpenVPN (GUI), macOS Tunnelblick, Linux OpenVPN (systemd managed), Linux OpenVPN (NM managed, CLI), Linux OpenVPN (NM managed, GUIs in various desktop environments).
> I think we did not revisit the dynamic challenge as few of those clients support it.
> The current implementation to prompt for TOTP - using either the dynamic (CRV1) or static (SCRV1) challenge response - enabled compatibility with all of those clients, which was crucial for migrating InfraVPN to SUSEID.
> 
> It sounds like a nice f(ea|u)ture idea to revisit AUTH_PENDING as an additional option, though.
